src/components/utils/controlled_or_internal_state.ts

```diff
@@ -0,0 +1,49 @@
+import { useCallback, useEffect, useRef, useState } from 'react';
+
+function usePrevious<T>(value: T) {
+  const ref = useRef<T>(value);
+
+  useEffect(() => {
+    ref.current = value;
+  }, [value]);
+
+  return ref.current;
```

Comment:
This is forbidden by React (and I think incompatible with React 19 or the React Compiler).

Reply:
I removed that part after using the suggestion radix-ui hook.